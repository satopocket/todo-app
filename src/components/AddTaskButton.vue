<template>
  <button @click="addTask()" class="button">Button</button>
</template>

<script lang="ts">
import Vue from "vue";
export default Vue.extend({
  name: "AddTaskButton",
  methods: {
    addTask(): void {
      this.$emit("add-task");
    }
  }
});
</script>

<style lang="scss" scoped>
.button {
  display: inline-block;
  padding: 0.3em 1em;
  text-decoration: none;
  color: #67c5ff;
  border: solid 2px #67c5ff;
  border-radius: 3px;
  transition: 0.4s;
  margin: 40px 0px 10px 20px;

  &:hover {
    background: #67c5ff;
    color: white;
  }
}
</style>
