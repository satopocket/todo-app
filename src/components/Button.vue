<template>
  <button @click.prevent="addTask()" class="button">Button</button>
</template>

<script lang="ts">
import Vue, { PropType } from "vue";
interface Task {
  name: string;
  done: boolean;
}
export default Vue.extend({
  name: "Button",
  methods: {
    addTask(): void {
      this.$emit("add-task");
    }
  }
});
</script>

<style lang="scss" scoped>
.button {
  display: inline-block;
  padding: 0.3em 1em;
  text-decoration: none;
  color: #67c5ff;
  border: solid 2px #67c5ff;
  border-radius: 3px;
  transition: 0.4s;
  margin: 40px 0px 10px 20px;

  &:hover {
    background: #67c5ff;
    color: white;
  }
}
</style>
