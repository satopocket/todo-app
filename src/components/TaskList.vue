<template>
  <div>
    <select v-model="isShownAllTask" class="select-box">
      <option :value="true">タスクをすべて表示</option>
      <option :value="false">完了済みのタスクを表示しない</option>
    </select>
    <div class="task-list">
      <template v-for="task in taskList">
        <label
          v-if="isShow(task.done)"
          class="task-list__item"
          :key="task.name"
        >
          <input type="checkbox" v-model="task.done" />
          {{ task.name }}
          <br />
        </label>
      </template>
      <p class="task-count" v-cloak>
        {{ showTaskCount }}/{{ taskList.length }}
      </p>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { PropType } from "vue";
import { Task } from "./types";
export default Vue.extend({
  name: "TaskList",
  data() {
    return {
      isShownAllTask: true
    } as {
      isShownAllTask: boolean;
    };
  },
  props: {
    taskList: {
      type: Array as PropType<Task[]>,
      default: () => [],
      required: true
    }
  },
  computed: {
    showTaskCount(): number {
      if (this.isShownAllTask) {
        return this.taskList.length;
      }
      return this.getUnfinishedNum;
    },
    getUnfinishedNum(): number {
      const unfinishedTasks = this.taskList.filter(task => !task.done);
      return unfinishedTasks.length;
    }
  },
  methods: {
    isShow(done: boolean): boolean {
      return this.isShownAllTask || (!this.isShownAllTask && !done);
    }
  }
});
</script>

<style lang="scss" scoped>
.select-box {
  position: absolute;
  top: 110px;
  left: 250px;
}

.task-list {
  margin-top: 50px;
  margin-left: 10px;
}

.task-count {
  margin-top: 40px;
  margin-left: 370px;
}
</style>
