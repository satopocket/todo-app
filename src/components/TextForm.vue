<template>
  <input v-model="inputValue" type="text" placeholder="タスク" class="input" />
</template>

<script lang="ts">
import Vue, { PropType } from "vue";
export default Vue.extend({
  name: "TextForm",
  model: {
    prop: "newTask",
    event: "input"
  },
  props: {
    newTask: {
      type: String as PropType<string>,
      default: ""
    }
  },
  computed: {
    inputValue: {
      get(): string {
        return this.newTask;
      },
      set(newVal: string) {
        if (this.newTask !== newVal) this.$emit("input", newVal);
      }
    }
  }
});
</script>

<style lang="scss" scoped>
.input {
  border: 1px solid;
  margin: 40px 0px 10px 10px;
}
</style>
