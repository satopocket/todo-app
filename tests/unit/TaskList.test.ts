import { shallowMount } from "@vue/test-utils";
import TaskList from "@/components/TaskList.vue";
import { Task } from "@/components/types";

const factory = () => {
  return shallowMount(TaskList, {
    propsData: {
      taskList: []
    }
  });
};

describe("TaskList.vue", () => {
  describe("snapshot", () => {
    test("タスクなし", () => {
      const wrapper = factory();
      expect(wrapper.element).toMatchSnapshot();
    });
    test("タスク1つ", () => {
      const wrapper = factory();
      const taskList: Task[] = [{ name: "task1", done: false }];
      wrapper.setProps({ taskList });
      expect(wrapper.element).toMatchSnapshot();
    });
    test("タスク2つ", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: false },
        { name: "task2", done: false }
      ];
      wrapper.setProps({ taskList });
      expect(wrapper.element).toMatchSnapshot();
    });
    test("タスク2つ中1つ完了", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: false },
        { name: "task2", done: true }
      ];
      wrapper.setProps({ taskList });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe("computed", () => {
    test("showTaskCount(isShownAllTask=true)", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: false },
        { name: "task2", done: false }
      ];
      wrapper.setProps({ taskList });
      const wm: any = wrapper.vm;
      wm.isShownAllTask = true;
      expect(wm.showTaskCount).toBe(2);
    });
    test("showTaskCount(isShownAllTask=false)", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: false },
        { name: "task2", done: true }
      ];
      wrapper.setProps({ taskList });
      const wm: any = wrapper.vm;
      wm.isShownAllTask = false;
      expect(wm.showTaskCount).toBe(wm.getUnfinishedNum);
    });
    test("getUnfinishedNum(未完了0件)", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: true },
        { name: "task2", done: true }
      ];
      wrapper.setProps({ taskList });
      const wm: any = wrapper.vm;
      expect(wm.getUnfinishedNum).toBe(0);
    });
    test("getUnfinishedNum(未完了1件)", () => {
      const wrapper = factory();
      const taskList: Task[] = [
        { name: "task1", done: false },
        { name: "task2", done: true }
      ];
      wrapper.setProps({ taskList });
      const wm: any = wrapper.vm;
      expect(wm.getUnfinishedNum).toBe(1);
    });
  });
  describe("methods", () => {
    test("isShow(isShownAllTask=true, done=true)", () => {
      const wrapper = factory();
      const wm: any = wrapper.vm;
      wm.isShownAllTask = true;
      expect(wm.isShow(true)).toBeTruthy();
    });
    test("isShow(isShownAllTask=true, done=false)", () => {
      const wrapper = factory();
      const wm: any = wrapper.vm;
      wm.isShownAllTask = true;
      expect(wm.isShow(false)).toBeTruthy();
    });
    test("isShow(isShownAllTask=false, done=true)", () => {
      const wrapper = factory();
      const wm: any = wrapper.vm;
      wm.isShownAllTask = false;
      expect(wm.isShow(true)).toBeFalsy();
    });
    test("isShow(isShownAllTask=false, done=false)", () => {
      const wrapper = factory();
      const wm: any = wrapper.vm;
      wm.isShownAllTask = false;
      expect(wm.isShow(false)).toBeTruthy();
    });
  });
});
