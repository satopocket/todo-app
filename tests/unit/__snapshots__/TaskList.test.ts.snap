// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList.vue snapshot タスク1つ 1`] = `
<div>
  <select
    class="select-box"
  >
    <option
      value="true"
    >
      タスクをすべて表示
    </option>
     
    <option
      value="false"
    >
      完了済みのタスクを表示しない
    </option>
  </select>
   
  <div
    class="task-list"
  >
    <label
      class="task-list__item"
    >
      <input
        type="checkbox"
      />
      
        task1
        
      <br />
    </label>
     
    <p
      class="task-count"
    >
      
      1/1
    
    </p>
  </div>
</div>
`;

exports[`TaskList.vue snapshot タスク2つ 1`] = `
<div>
  <select
    class="select-box"
  >
    <option
      value="true"
    >
      タスクをすべて表示
    </option>
     
    <option
      value="false"
    >
      完了済みのタスクを表示しない
    </option>
  </select>
   
  <div
    class="task-list"
  >
    <label
      class="task-list__item"
    >
      <input
        type="checkbox"
      />
      
        task1
        
      <br />
    </label>
    <label
      class="task-list__item"
    >
      <input
        type="checkbox"
      />
      
        task2
        
      <br />
    </label>
     
    <p
      class="task-count"
    >
      
      2/2
    
    </p>
  </div>
</div>
`;

exports[`TaskList.vue snapshot タスク2つ中1つ完了 1`] = `
<div>
  <select
    class="select-box"
  >
    <option
      value="true"
    >
      タスクをすべて表示
    </option>
     
    <option
      value="false"
    >
      完了済みのタスクを表示しない
    </option>
  </select>
   
  <div
    class="task-list"
  >
    <label
      class="task-list__item"
    >
      <input
        type="checkbox"
      />
      
        task1
        
      <br />
    </label>
    <label
      class="task-list__item"
    >
      <input
        type="checkbox"
      />
      
        task2
        
      <br />
    </label>
     
    <p
      class="task-count"
    >
      
      2/2
    
    </p>
  </div>
</div>
`;

exports[`TaskList.vue snapshot タスクなし 1`] = `
<div>
  <select
    class="select-box"
  >
    <option
      value="true"
    >
      タスクをすべて表示
    </option>
     
    <option
      value="false"
    >
      完了済みのタスクを表示しない
    </option>
  </select>
   
  <div
    class="task-list"
  >
     
    <p
      class="task-count"
    >
      
      0/0
    
    </p>
  </div>
</div>
`;
